File: CSharp/GetRealGoodUtils.Tests/StringsTests.cs
using System;
using NUnit.Framework;
using System.Collections.Generic;

namespace GetRealGood.Utils.Tests
{
    public class StringsTests
    {

        [OneTimeSetUp]
        public static void SetupAll()
        {
        }

        [OneTimeTearDown]
        public static void TearDownAll()
        {
        }

        [SetUp]
        public void Setup()
        {
        }

        [TearDown]
        public void TearDown()
        {
        }

        class ToUpperTests
        {
            [SetUp]
            public void Setup()
            {
            }

            [TearDown]
            public void TearDown()
            {
            }

            [Test]
            public void ToUpper_Common_String_String()
            {

                try
                {

                    string s = "Hello, World!";

                    string result = Strings.ToUpper(s);
                    Assert.AreEqual("HELLO, WORLD!", result);
                }
                catch (Exception ex)
                {
                    Assert.Fail(ex.Message);
                }

            }

            [Test]
            public void ToUpper_Nullable_Null_StringsException()
            {

                try
                {
                    string s = null;
                    Assert.Throws(typeof(StringsException), delegate { Strings.ToUpper(s); });

                }
                catch (Exception ex)
                {
                    Assert.Fail(ex.Message);
                }

            }

            [Test]
            public void ToUpper_Empty_String_String()
            {

                try
                {
                    string s = "";
                    Assert.AreEqual(s, Strings.ToUpper(s));

                }
                catch (Exception ex)
                {
                    Assert.Fail(ex.Message);
                }

            }

            [Test]
            public void ToUpper_Exceptional_String_StringsException()
            {

                try
                {
                    String s = null;
                    Assert.Throws(typeof(StringsException), delegate { Strings.ToUpper(s); });

                }
                catch (Exception ex)
                {
                    Assert.Fail(ex.Message);
                }

            }
        }

        class ToLowerTests
        {

            [SetUp]
            public void Setup()
            {
            }

            [TearDown]
            public void TearDown()
            {
            }

            [Test]
            public void ToLower_Common_String_String()
            {

                try
                {
                    string s = "Hello, World!";
                    s = Strings.ToLower(s);
                    Assert.AreNotEqual(null, s);
                    Assert.AreEqual("hello, world!", s);
                }
                catch (Exception ex)
                {
                    Assert.Fail(ex.Message);
                }

            }

            [Test]
            public void ToLower_Nullable_Null_StringsException()
            {

                try
                {
                    String s = null;
                    Assert.Throws(typeof(StringsException), delegate { Strings.ToLower(s); });
                    // Assert.That(delegate { Strings.ToLower(s); }, Throws.Exception.TypeOf(typeof(StringsException)));
                    

                }
                catch (Exception ex)
                {
                    Assert.Fail(ex.Message);

                }
            }

            [Test]
            public void ToLower_Empty_String_String()
            {

                try
                {
                    string s = "";
                    Assert.AreEqual(s, Strings.ToLower(s));

                }
                catch (Exception ex)
                {
                    Assert.Fail(ex.Message);

                }
            }

            [Test]
            public void ToLower_Exceptional_String_StringsException()
            {

                try
                {
                    string s = null;
                    Assert.Throws(typeof(StringsException), delegate { Strings.ToLower(s); });

                }
                catch (Exception ex)
                {
                    Assert.Fail(ex.Message);

                }
            }
        }

        
         
    class SplitTests {

        [SetUp]
        public void SetUp() {

        }

        [TearDown]
        public void TearDown() {

        }

        [Test]
        public void Split_Common_String_Boolean_String() {
            try {
                String s = "Hello, World!";
                var result = Strings.Split(s, ',', false);
                Assert.AreEqual(result.Contains("Hello"), true);
                Assert.AreEqual(result.Contains(" World!"), true);
                Assert.AreEqual(result[0], "Hello");
                Assert.AreEqual(result[1], " World!");

            } catch (Exception ex) {
                Assert.Fail(ex.Message);

            }
        }

        [Test]
        public void Split_Common_String_Boolean_String__1() {
            try {
                String s = "Hello, World! ";
                var result = Strings.Split(s, ',', true);
                Assert.AreEqual(result.Contains("Hello"), true);
                Assert.AreEqual(result.Contains("World!"), true);
                Assert.AreEqual(result[0], "Hello");
                Assert.AreEqual(result[1], "World!");

            } catch (Exception ex) {
                Assert.Fail(ex.Message);

            }
        }

        [Test]
        public void Split_Common_String_Boolean_String__2() {
            try {
                String s = "Hello, World! ";
                var result = Strings.Split(s, ',', false);
                Assert.AreEqual(result.Contains("Hello"), true);
                Assert.AreEqual(result.Contains(" World! "), true);
                Assert.AreEqual(result[0], "Hello");
                Assert.AreEqual(result[1], " World! ");

            } catch (Exception ex) {
                Assert.Fail(ex.Message);

            }
        }

        [Test]
        public void Split_Nullable_Null_Boolean_StringsException() {
            try {
                String s = null;
                Assert.Throws(typeof(StringsException), delegate {
                    Strings.Split(s, ',', false);
                });
            } catch (Exception ex) {
                Assert.Fail(ex.Message);

            }
        }

        [Test]
        public void Split_Empty_String_Boolean_String() {
            try {
                String s = "";
                Assert.AreEqual(0, Strings.Split(s, '`', false).Count);

            } catch (Exception ex) {
                Assert.Fail(ex.Message);

            }
        }
    }

        class ConcatTests
        {

            [SetUp]
            public void Setup()
            {
            }

            [TearDown]
            public void TearDown()
            {
            }


            [Test]
            public void Concat_Common_StringList_String_Boolean_String()
            {
                try
                {
                    IList<string> s = new List<string> {"Hello", "World!"};
                    var result = Strings.Concat(s, ", ", true);
                    Assert.AreEqual("Hello, World!", result);
                }
                catch (Exception ex)
                {
                    Assert.Fail(ex.Message);

                }
            }

            [Test]
            public void Concat_Common_StringList_String_Boolean_String__1()
            {
                try
                {
                    IList<string> s = new List<string> {"Hello ", " World!"};
                    var result = Strings.Concat(s, ", ", false);
                    Assert.AreEqual("Hello ,  World!", result);
                }
                catch (Exception ex)
                {
                    Assert.Fail(ex.Message);

                }
            }

            [Test]
            public void Concat_Common_StringList_String_Boolean_String__2()
            {
                try
                {
                    IList<string> s = new List<string> {"Hello ", " World!"};
                    var result = Strings.Concat(s, ", ", true);
                    Assert.AreEqual("Hello, World!", result);
                }
                catch (Exception ex)
                {
                    Assert.Fail(ex.Message);

                }
            }


            [Test]
            public void Concat_Empty_StringList_String_Boolean_String()
            {
                try
                {
                    IList<string> s = new List<string> {"", ""};
                    var result = Strings.Concat(s, ", ", false);
                    Assert.AreEqual("", result);
                }
                catch (Exception ex)
                {
                    Assert.Fail(ex.Message);

                }
            }

            [Test]
            public void Concat_Empty_StringList_String_Boolean_String__1()
            {
                try
                {
                    IList<string> s = new List<string> {"", ""};
                    var result = Strings.Concat(s, ", ", true);
                    Assert.AreEqual("", result);
                }
                catch (Exception ex)
                {
                    Assert.Fail(ex.Message);

                }
            }

            [Test]
            public void Concat_Nullable_Null_String_Boolean_StringsException()
            {
                try
                {
                    IList<string> s = null;
                    Assert.Throws(typeof(StringsException), delegate { Strings.Concat(s, ",", true); });
                }
                catch (Exception ex)
                {
                    Assert.Fail(ex.Message);

                }
            }


            [Test]
            public void Concat_Nullable_Null_String_Boolean_StringsException__1()
            {
                try
                {
                    IList<String> s = null;
                    Assert.Throws(typeof(StringsException), delegate { Strings.Concat(s, ",", false); });
                }
                catch (Exception ex)
                {
                    Assert.Fail(ex.Message);

                }
            }
        }

        class ReplaceTests
        {

            [SetUp]
            public void Setup()
            {
            }

            [TearDown]
            public void TearDown()
            {
            }

            [Test]
            public void Replace_Common_String_Char_Char_Boolean_String()
            {

                try
                {

                    string origin = "Hello, World! ";
                    char original = 'o';
                    char replace = 'e';
                    string result = Strings.Replace(origin, original, replace, true);
                    Assert.AreEqual("Helle, Werld!", result);

                }
                catch (Exception ex)
                {
                    Assert.Fail(ex.Message);
                }
            }

            [Test]
            public void Replace_Common_String_Char_Char_Boolean_String__1()
            {

                try
                {

                    string origin = "Hello, World! ";
                    char original = 'o';
                    char replace = 'e';
                    string result = Strings.Replace(origin, original, replace, false);
                    Assert.AreEqual("Helle, Werld! ", result);

                }
                catch (Exception ex)
                {
                    Assert.Fail(ex.Message);
                }
            }

            [Test]
            public void Replace_Empty_String_Char_Char_Boolean_String()
            {

                try
                {

                    string origin = "";
                    char original = 'o';
                    char replace = 'e';
                    string result = Strings.Replace(origin, original, replace, true);
                    Assert.AreEqual("", result);

                }
                catch (Exception ex)
                {
                    Assert.Fail(ex.Message);
                }
            }

            [Test]
            public void Replace_Empty_String_Char_Char_Boolean_String__1()
            {

                try
                {

                    string origin = "";
                    char original = 'o';
                    char replace = 'e';
                    string result = Strings.Replace(origin, original, replace, false);
                    Assert.AreEqual("", result);

                }
                catch (Exception ex)
                {
                    Assert.Fail(ex.Message);
                }
            }

            [Test]
            public void Replace_Nullable_Null_Char_Char_Boolean_StringsException()
            {

                try
                {

                    string origin = null;
                    char original = 'o';
                    char replace = 'e';
                    Assert.Throws(typeof(StringsException), delegate { Strings.Replace(origin, 'a', 'b', true); });

                }
                catch (Exception ex)
                {
                    Assert.Fail(ex.Message);
                }
            }

            [Test]
            public void Replace_Nullable_Null_Char_Char_Boolean_StringsException__1()
            {

                try
                {

                    string origin = null;
                    char original = 'o';
                    char replace = 'e';
                    Assert.Throws(typeof(StringsException), delegate { Strings.Replace(origin, 'a', 'b', true); });
                }
                catch (Exception ex)
                {
                    Assert.Fail(ex.Message);
                }
            }

            [Test]
            public void Replace_Special_String_Char_Char_Boolean_String()
            {

                try
                {

                    string origin = "Hello, World! ";
                    char original = 'x';
                    char replace = 'e';
                    string result = Strings.Replace(origin, original, replace, true);
                    Assert.AreEqual("Hello, World!", result);

                }
                catch (Exception ex)
                {
                    Assert.Fail(ex.Message);
                }
            }

            [Test]
            public void Replace_Special_String_Char_Char_Boolean_String__1()
            {

                try
                {

                    string origin = "Hello, World! ";
                    char original = 'x';
                    char replace = 'e';
                    string result = Strings.Replace(origin, original, replace, false);
                    Assert.AreEqual("Hello, World! ", result);

                }
                catch (Exception ex)
                {
                    Assert.Fail(ex.Message);
                }
            }

        }

        class ReplaceTests__1
        {

            [SetUp]
            public void Setup()
            {
            }

            [TearDown]
            public void TearDown()
            {
            }

            
            [Test]
            public void Replace_Common_String_Char_Char_String()
            {

                try
                {

                    string origin = "Hello, World! ";
                    char original = 'o';
                    char replace = 'e';
                    string result = Strings.Replace(origin, original, replace);
                    Assert.AreEqual("Helle, Werld! ", result);

                }
                catch (Exception ex)
                {
                    Assert.Fail(ex.Message);
                }
            }

            [Test]
            public void Replace_Empty_String_Char_Char_String()
            {

                try
                {

                    string origin = "";
                    char original = 'o';
                    char replace = 'e';
                    string result = Strings.Replace(origin, original, replace);
                    Assert.AreEqual("", result);

                }
                catch (Exception ex)
                {
                    Assert.Fail(ex.Message);
                }
            }

            [Test]
            public void Replace_Nullable_Null_Char_Char_Boolean_StringsException()
            {

                try
                {

                    string origin = null;
                    char original = 'o';
                    char replace = 'e';
                    Assert.Throws(typeof(StringsException), delegate { Strings.Replace(origin, original, replace); });

                }
                catch (Exception ex)
                {
                    Assert.Fail(ex.Message);
                }
            }
        }
    }

}

File: Java/test/tech/getrealgood/util/test/StringsTests.java
package tech.getrealgood.util.test;

import org.junit.jupiter.api.*;
import org.junit.platform.commons.logging.Logger;
import org.junit.platform.commons.logging.LoggerFactory;
import tech.getrealgood.util.Strings;
import tech.getrealgood.util.StringsException;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import static org.junit.jupiter.api.Assertions.*;

class StringsTests {

    private static final Logger log = LoggerFactory.getLogger(StringsTests.class);

    @Nested
    class ToUpperTests {

        @Test
        void toUpper_Common_String_String() {

            try {

                String s = "Hello, World!";

                String result = Strings.toUpper(s);
                assertEquals("HELLO, WORLD!", result);
            } catch (Exception ex) {
                fail(ex.getMessage());
            }

        }

        @Test
        void toUpper_Common_String_String__1() {

            try {

                String s = "Hello, World!";

                String result = Strings.toUpper(s);
                assertNotEquals(s, result);
            } catch (Exception ex) {
                fail(ex.getMessage());
            }

        }

        @Test
        void toUpper_Common_String_String__2() {

            try {

                String s = "Hello, World!";

                String result = Strings.toUpper(s);
                assertTrue(!s.equals(result));
            } catch (Exception ex) {
                fail(ex.getMessage());
            }

        }

        @Test
        void toUpper_Common_String_String__3() {

            try {

                String s = "Hello, World!";

                String result = Strings.toUpper(s);
                assertFalse(s.equals(result));
            } catch (Exception ex) {
                fail(ex.getMessage());
            }

        }

        @Test
        void toUpper_Common_String_String__4() {

            try {

                String s = "Hello, World!";

                String result = Strings.toUpper(s);
                assertNotNull(result);
            } catch (Exception ex) {
                fail(ex.getMessage());
            }

        }

        @Test
        void toUpper_Common_String_String__5() {

            try {

                String s = "Hello, World!";

                assertDoesNotThrow(() -> {
                    Strings.toUpper(s);
                });
            } catch (Exception ex) {
                fail(ex.getMessage());
            }

        }

        @Test
        void toUpper_Nullable_Null_StringsException() {

            try {
                String s = null;
                assertThrows(StringsException.class, () -> {
                    Strings.toUpper(s);
                });

            } catch (Exception ex) {
                fail(ex.getMessage());
            }

        }

        @Test
        void toUpper_Empty_String_String() {

            try {
                String s = "";
                assertEquals(s, Strings.toUpper(s));

            } catch (Exception ex) {
                fail(ex.getMessage());
            }

        }

        @Test
        void toUpper_Exceptional_String_StringsException() {

            try {
                String s = null;
                assertThrows(StringsException.class, () -> {
                    Strings.toUpper(s);
                });

            } catch (Exception ex) {
                fail(ex.getMessage());
            }

        }

    }

    @Nested
    class ToUpperTests__1 {

        @Test
        void toUpper_Common_StringList_StringList() {

            try {

                var s = new ArrayList(3);

                s.add("Hello, World!");
                s.add("hello, world!");
                s.add("HELLO, WORLD!");

                var s2 = new ArrayList(3);

                s2.add("HELLO, WORLD!");
                s2.add("HELLO, WORLD!");
                s2.add("HELLO, WORLD!");

                var result = Strings.toUpper(s);

                assertLinesMatch(s2, result);
            } catch (Exception ex) {
                fail(ex.getMessage());
            }

        }

        @Test
        void toUpper_Common_StringList_StringList__1() {

            try {

                var s = new String[]{"Hello, World!", "hello, world!", "HELLO, WORLD!"};

                var s2 = new String[]{"HELLO, WORLD!", "HELLO, WORLD!", "HELLO, WORLD!"};

                var result = Strings.toUpper(Arrays.asList(s));

                assertArrayEquals(s2, result.toArray());
            } catch (Exception ex) {
                fail(ex.getMessage());
            }

        }

        @Test
        void toUpper_Nullable_Null_StringsException() {

            try {
                List s = null;
                assertThrows(StringsException.class, () -> {
                    Strings.toUpper(s);
                });

            } catch (Exception ex) {
                fail(ex.getMessage());
            }

        }

        @Test
        void toUpper_Empty_StringList_String() {

            try {
                List s = new ArrayList<>(0);
                assertEquals(s, Strings.toUpper(s));

            } catch (Exception ex) {
                fail(ex.getMessage());
            }

        }

        @Test
        void toUpper_Exceptional_StringList_StringsException() {

            try {
                var s = new ArrayList(3);

                s.add("Hello, World!");
                s.add(null);
                s.add("HELLO, WORLD!");

                assertThrows(StringsException.class, () -> {
                    Strings.toUpper(s);
                });

            } catch (Exception ex) {
                fail(ex.getMessage());
            }

        }

    }


    @Nested
    class ToLowerTests {

        @Test
        void toLower_Common_String_String() {

            try {
                String s = "Hello, World!";
                s = Strings.toLower(s);
                assertNotEquals(null, s);
                assertEquals("hello, world!", s);
            } catch (Exception ex) {
                fail(ex.getMessage());
            }

        }

        @Test
        void toLower_Nullable_Null_StringsException() {

            try {
                String s = null;
                assertThrows(StringsException.class, () -> {
                    Strings.toLower(s);
                });

            } catch (Exception ex) {
                fail(ex.getMessage());

            }
        }

        @Test
        void toLower_Empty_String_String() {

            try {
                String s = "";
                assertEquals(s, Strings.toLower(s));

            } catch (Exception ex) {
                fail(ex.getMessage());

            }
        }

        @Test
        void toLower_Exceptional_String_StringsException() {

            try {
                String s = null;
                assertThrows(StringsException.class, () -> {
                    Strings.toLower(s);
                });

            } catch (Exception ex) {
                fail(ex.getMessage());

            }
        }
    }

    @Nested
    class ToLowerTests__1 {

        @Test
        void toLower_Common_StringList_StringList() {

            try {

                var s = new ArrayList(3);

                s.add("Hello, World!");
                s.add("hello, world!");
                s.add("HELLO, WORLD!");

                var s2 = new ArrayList(3);

                s2.add("hello, world!");
                s2.add("hello, world!");
                s2.add("hello, world!");

                var result = Strings.toLower(s);

                assertLinesMatch(s2, result);
            } catch (Exception ex) {
                fail(ex.getMessage());
            }

        }

        @Test
        void toLower_Common_StringList_StringList__1() {

            try {

                var s = new ArrayList(3);

                s.add("Hello, World!");
                s.add("hello, world!");
                s.add("HELLO, WORLD!");

                var s2 = new ArrayList(3);

                s2.add("hello, world!");
                s2.add("hello, world!");
                s2.add("hello, world!");

                var result = Strings.toLower(s);

                assertIterableEquals(s2, result);
            } catch (Exception ex) {
                fail(ex.getMessage());
            }

        }

        @Test
        void toLower_Common_StringList_StringList__2() {

            try {

                var s = new ArrayList(3);

                s.add("Hello, World!");
                s.add("hellO, world!");
                s.add("HELLO, WORLD!");

                var s2 = new ArrayList(3);

                s2.add("hello, world!");
                s2.add("hello, world!");
                s2.add("hello, world!");

                var result = Strings.toLower(s);

                for (var i = 0; i < s2.size(); i++) {
                    assertNotSame(s2.get(i), result.get(i));
                }

            } catch (Exception ex) {
                fail(ex.getMessage());
            }

        }

        @Test
        void toLower_Common_StringList_StringList__3() {

            try {

                var s = new ArrayList(3);

                s.add("hello, world!");
                s.add("hello, world!");
                s.add("hello, world!");

                var s2 = new ArrayList(3);

                s2.add("hello, world!");
                s2.add("hello, world!");
                s2.add("hello, world!");

                var result = Strings.toLower(s);

                for (var i = 0; i < s2.size(); i++) {
                    assertSame(s2.get(i), result.get(i));
                }

            } catch (Exception ex) {
                fail(ex.getMessage());
            }

        }


        @Test
        void toLower_Common_StringList_StringList__4() {

            try {

                var s = new String[]{"Hello, World!", "hello, world!", "HELLO, WORLD!"};

                var s2 = new String[]{"hello, world!", "hello, world!", "hello, world!"};

                var result = Strings.toLower(Arrays.asList(s));

                assertArrayEquals(s2, result.toArray());
            } catch (Exception ex) {
                fail(ex.getMessage());
            }

        }


        @Test
        void toLower_Nullable_Null_StringsException() {

            try {
                List s = null;
                assertThrows(StringsException.class, () -> {
                    Strings.toLower(s);
                });

            } catch (Exception ex) {
                fail(ex.getMessage());
            }

        }

        @Test
        void toLower_Empty_StringList_String() {

            try {
                List s = new ArrayList<>(0);
                assertEquals(s, Strings.toLower(s));

            } catch (Exception ex) {
                fail(ex.getMessage());
            }

        }

        @Test
        void toLower_Exceptional_StringList_StringsException() {

            try {
                var s = new ArrayList(3);

                s.add("Hello, World!");
                s.add(null);
                s.add("HELLO, WORLD!");

                assertThrows(StringsException.class, () -> {
                    Strings.toLower(s);
                });

            } catch (Exception ex) {
                fail(ex.getMessage());
            }

        }

    }


    @Nested
    class SplitTests {

        @Test
        void split_Common_String_Boolean_String() {
            try {
                String s = "Hello, World!";
                var result = Strings.split(s, ',', false);
                assertEquals(true, result.contains("Hello"));
                assertEquals(true, result.contains(" World!"));
                assertEquals("Hello", result.get(0));
                assertEquals(" World!", result.get(1));

            } catch (Exception ex) {
                fail(ex.getMessage());

            }
        }

        @Test
        void split_Common_String_Boolean_String__1() {
            try {
                String s = "Hello, World! ";
                var result = Strings.split(s, ',', true);
                assertEquals(true, result.contains("Hello"));
                assertEquals(true, result.contains("World!"));
                assertEquals("Hello", result.get(0));
                assertEquals("World!", result.get(1));

            } catch (Exception ex) {
                fail(ex.getMessage());

            }
        }

        @Test
        void split_Common_String_Boolean_String__2() {
            try {
                String s = "Hello, World! ";
                var result = Strings.split(s, ',', false);
                assertEquals(true, result.contains("Hello"));
                assertEquals(true, result.contains(" World! "));
                assertEquals("Hello", result.get(0));
                assertEquals(" World! ", result.get(1));

            } catch (Exception ex) {
                fail(ex.getMessage());

            }
        }

        @Test
        void split_Nullable_Null_Boolean_StringsException() {
            try {
                String s = null;
                assertThrows(StringsException.class, () -> {
                    Strings.split(s, ',', false);
                });
            } catch (Exception ex) {
                fail(ex.getMessage());

            }
        }

        @Test
        void split_Empty_String_Boolean_String() {
            try {
                String s = "";
                assertEquals(0, Strings.split(s, '`', false).size());

            } catch (Exception ex) {
                fail(ex.getMessage());

            }
        }
    }

    @Nested
    class ConcatTests {

        @Test
        void concat_Common_StringList_String_Boolean_String() {
            try {
                List s = Arrays.asList("Hello", "World!");
                var result = Strings.concat(s, ", ", true);
                assertEquals("Hello, World!", result);
            } catch (Exception ex) {
                fail(ex.getMessage());

            }
        }

        @Test
        void concat_Common_StringList_String_Boolean_String__1() {
            try {
                List s = Arrays.asList("Hello ", " World!");
                var result = Strings.concat(s, ", ", false);
                assertEquals("Hello ,  World!", result);
            } catch (Exception ex) {
                fail(ex.getMessage());

            }
        }

        @Test
        void concat_Common_StringList_String_Boolean_String__2() {
            try {
                List s = Arrays.asList("Hello ", " World!");
                var result = Strings.concat(s, ", ", true);
                assertEquals("Hello, World!", result);
            } catch (Exception ex) {
                fail(ex.getMessage());

            }
        }


        @Test
        void concat_Empty_StringList_String_Boolean_String() {
            try {
                List s = Arrays.asList("", "");
                var result = Strings.concat(s, ", ", false);
                assertEquals("", result);
            } catch (Exception ex) {
                fail(ex.getMessage());

            }
        }

        @Test
        void concat_Empty_StringList_String_Boolean_String__1() {
            try {
                List s = Arrays.asList("", "");
                var result = Strings.concat(s, ", ", true);
                assertEquals("", result);
            } catch (Exception ex) {
                fail(ex.getMessage());

            }
        }

        @Test
        void concat_Nullable_Null_String_Boolean_StringsException() {
            try {
                List s = null;
                assertThrows(StringsException.class, () -> {
                    Strings.concat(s, ",", true);
                });
            } catch (Exception ex) {
                fail(ex.getMessage());

            }
        }


        @Test
        void concat_Nullable_Null_String_Boolean_StringsException__1() {
            try {
                List s = null;
                assertThrows(StringsException.class, () -> {
                    Strings.concat(s, ",", false);
                });
            } catch (Exception ex) {
                fail(ex.getMessage());

            }
        }
    }

    @Nested
    class ReplaceTests {

        @Test
        public void replace_Common_String_Char_Char_Boolean_String() {

            try {

                String origin = "Hello, World! ";
                char original = 'o';
                char replace = 'e';
                String result = Strings.replace(origin, original, replace, true);
                assertEquals("Helle, Werld!", result);

            } catch (Exception ex) {
                fail(ex.getMessage());
            }
        }

        @Test
        public void replace_Common_String_Char_Char_Boolean_String__1() {

            try {

                String origin = "Hello, World! ";
                char original = 'o';
                char replace = 'e';
                String result = Strings.replace(origin, original, replace, false);
                assertEquals("Helle, Werld! ", result);

            } catch (Exception ex) {
                fail(ex.getMessage());
            }
        }

        @Test
        public void replace_Empty_String_Char_Char_Boolean_String() {

            try {

                String origin = "";
                char original = 'o';
                char replace = 'e';
                String result = Strings.replace(origin, original, replace, true);
                assertEquals("", result);

            } catch (Exception ex) {
                fail(ex.getMessage());
            }
        }

        @Test
        public void replace_Empty_String_Char_Char_Boolean_String__1() {

            try {

                String origin = "";
                char original = 'o';
                char replace = 'e';
                String result = Strings.replace(origin, original, replace, false);
                assertEquals("", result);

            } catch (Exception ex) {
                fail(ex.getMessage());
            }
        }

        @Test
        public void replace_Nullable_Null_Char_Char_Boolean_StringsException() {

            try {

                String origin = null;
                char original = 'o';
                char replace = 'e';
                assertThrows(StringsException.class, () -> {
                    Strings.replace(origin, 'a', 'b', true);
                });

            } catch (Exception ex) {
                fail(ex.getMessage());
            }
        }

        @Test
        public void replace_Nullable_Null_Char_Char_Boolean_StringsException__1() {

            try {

                String origin = null;
                char original = 'o';
                char replace = 'e';
                assertThrows(StringsException.class, () -> {
                    Strings.replace(origin, 'a', 'b', true);
                });
            } catch (Exception ex) {
                fail(ex.getMessage());
            }
        }

        @Test
        public void replace_Special_String_Char_Char_Boolean_String() {

            try {

                String origin = "Hello, World! ";
                char original = 'x';
                char replace = 'e';
                String result = Strings.replace(origin, original, replace, true);
                assertEquals("Hello, World!", result);

            } catch (Exception ex) {
                fail(ex.getMessage());
            }
        }

        @Test
        public void replace_Special_String_Char_Char_Boolean_String__1() {

            try {

                String origin = "Hello, World! ";
                char original = 'x';
                char replace = 'e';
                String result = Strings.replace(origin, original, replace, false);
                assertEquals("Hello, World! ", result);

            } catch (Exception ex) {
                fail(ex.getMessage());
            }
        }

    }

    @Nested
    class ReplaceTests__1 {

        @Test
        public void replace_Common_String_Char_Char_String() {

            try {

                String origin = "Hello, World! ";
                char original = 'o';
                char replace = 'e';
                String result = Strings.replace(origin, original, replace);
                assertEquals("Helle, Werld! ", result);

            } catch (Exception ex) {
                fail(ex.getMessage());
            }
        }

        @Test
        public void replace_Empty_String_Char_Char_String() {

            try {

                String origin = "";
                char original = 'o';
                char replace = 'e';
                String result = Strings.replace(origin, original, replace);
                assertEquals("", result);

            } catch (Exception ex) {
                fail(ex.getMessage());
            }
        }

        @Test
        public void replace_Nullable_Null_Char_Char_Boolean_StringsException() {

            try {

                String origin = null;
                char original = 'o';
                char replace = 'e';
                assertThrows(StringsException.class, () -> {
                    Strings.replace(origin, original, replace);
                });

            } catch (Exception ex) {
                fail(ex.getMessage());
            }
        }

    }
}